stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt

```diff
@@ -83,53 +85,56 @@ internal class ChannelControllerImpl(
-    private val _channelData = MutableLiveData<ChannelData>()
-    private val _oldMessages = MutableLiveData<Map<String, Message>>()
+    private val _messages = MutableStateFlow<Map<String, Message>?>(null)
+    private val _watcherCount = MutableStateFlow<Int>(0)
+    private val _typing = MutableStateFlow<Map<String, ChatEvent>>(emptyMap())
+    private val _reads = MutableStateFlow<Map<String, ChannelUserRead>>(emptyMap())
+    private val _read = MutableStateFlow<ChannelUserRead?>(null)
+    private val _endOfNewerMessages = MutableStateFlow(false)
+    private val _endOfOlderMessages = MutableStateFlow(false)
+    private val _loading = MutableStateFlow(false)
+    private val _hidden = MutableStateFlow(false)
+    private val _muted = MutableStateFlow(false)
+    private val _watchers = MutableStateFlow<Map<String, User>>(emptyMap())
+    private val _members = MutableStateFlow<Map<String, Member>>(emptyMap())
+    private val _loadingOlderMessages = MutableStateFlow(false)
+    private val _loadingNewerMessages = MutableStateFlow(false)
+    private val _channelData = MutableStateFlow<ChannelData?>(null)
+    private val _oldMessages = MutableStateFlow<Map<String, Message>?>(emptyMap())
     internal var hideMessagesBefore: Date? = null
-    val unfilteredMessages: LiveData<List<Message>> =
-        Transformations.map(_messages) { it.values.toList() }
+    val unfilteredMessages: LiveData<List<Message>?> = _messages.map {
+        if (it == null) { null } else { it.values.toList() }
+    }.asLiveData()
 
     /** a list of messages sorted by message.createdAt */
     override val messages: LiveData<List<Message>> = messagesTransformation(_messages)
 
     override val oldMessages: LiveData<List<Message>> = messagesTransformation(_oldMessages)
 
-    private fun messagesTransformation(messages: MutableLiveData<Map<String, Message>>): LiveData<List<Message>> {
-        return Transformations.map(messages) { messageMap ->
-            messageMap.values
-                .asSequence()
-                .filter { it.parentId == null || it.showInChannel }
-                .filter { hideMessagesBefore == null || it.wasCreatedAfter(hideMessagesBefore) }
-                .sortedBy { it.createdAt ?: it.createdLocallyAt }
-                .toList()
-                .map { it.copy() }
+    private fun messagesTransformation(messages: MutableStateFlow<Map<String, Message>?>): LiveData<List<Message>> {
+        return Transformations.map(messages.asLiveData()) { messageMap ->
+            if (messageMap == null) { null } else {
+                messageMap.values
+                    .asSequence()
+                    .filter { it.parentId == null || it.showInChannel }
+                    .filter { hideMessagesBefore == null || it.wasCreatedAfter(hideMessagesBefore) }
+                    .sortedBy { it.createdAt ?: it.createdLocallyAt }
+                    .toList()
+                    .map { it.copy() }
+            }
```

Comment:
The advantage of flow usage is flow operators. I'd suggest the next code for this snippet:

``` kotlin
/** a list of messages sorted by message.createdAt */
    override val messages: LiveData<List<Message>> = _messages.messagesTransformation().asLiveData()

    override val oldMessages: LiveData<List<Message>> =
        _oldMessages.messagesTransformation().asLiveData()

    private fun Flow<Map<String, Message>?>.messagesTransformation(): Flow<List<Message>> =
        map { it?.values ?: emptyList() }
            .map { it.asSequence() }
            .map { messages -> messages.filter { it.parentId == null || it.showInChannel } }
            .map { messages ->
                messages.filter {
                    hideMessagesBefore == null || it.wasCreatedAfter(hideMessagesBefore)
                }
            }
            .map { messages -> messages.sortedBy { it.createdAt ?: it.createdLocallyAt } }
            .map { it.toList() }
```

Reply:
Trying to keep the PR small, but yes that approach is easier/cleaner than livedata's transformation step

Reply:
I strongly agree, we should get rid of `LiveData` transformations as soon as we can, it's almost a shame that they were ever provided as a possible API to use, pretty misleading about what `LiveData` is supposed to be used for. I removed them in https://github.com/GetStream/stream-chat-android/pull/824/commits/8cd9b08fd4b6cd22b223f5a128802e6cfb85800b.

Reply:
We could also remove more livedata in the QueryChannel and thread controllers if we want to go for a full replace.

Reply:
We should eventually, but if we don't need it for bugfixing elsewhere (like we did here), then it's just a nice to have thing, which can wait for now.